File: problems/cpp/intersection-of-two-linked-lists.cc
// 32 ms, 09/01/2017
void advance_head(ListNode **head, int num)
{
    for (; num; *head = (*head)->next, --num);
}

int count_nodes(ListNode *head)
{
    int count = 0;
    for (; head; head = head->next, ++count);
    return count;
}

ListNode *getIntersectionNode(ListNode *head_a, ListNode *head_b)
{
    // Count the number of nodes in each list
    int count_a = count_nodes(head_a), count_b = count_nodes(head_b);
    
    // Bring head_a and head_b at the same distance from the end(s) 
    if (count_a > count_b)
        advance_head(&head_a, count_a - count_b);
    else if (count_b > count_a)
        advance_head(&head_b, count_b - count_a);

    // head_a and head_b are at the same distance from the end(s)
    while (head_a) {
        if (head_a == head_b)
            return head_a;
        head_a = head_a->next, head_b = head_b->next;
    }

    return nullptr;
}

File: code/python/intersection-of-two-linked-lists.py
# Date  : 03/25/2018
# Time  : O(n)
# Space : O(1)

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        def len(head):
            count = 0
            while head:
                count += 1
                head = head.next
            return count

        len1, len2 = len(headA), len(headB)

        # if listB is longer than listA, swap them
        if len2 > len1:
            headA, headB = headB, headA

        # Advance listA head so that it is at the same distance
        # from the end as listB head is from its end
        diff = abs(len1 - len2)
        for _ in range(diff):
            headA = headA.next

        # Now advance both list heads in tandem
        while headA:
            if headA is headB:
                break
            headA, headB = headA.next, headB.next

        return headA
